Construct `Payout` based on interval

luckysori opened the pull request:
Fixes #68.

This is now closer to what I think we should do when adding the feature of supporting an oracle/protocol combination with payouts based on price intervals (rather than concrete prices).

I will very likely end up squashing a lot of this, but I think this can still be reviewed in its current state **patch by patch** (probably too much going on overall). Particularly to @thomaseizinger, who reviewed it the first time, https://github.com/comit-network/hermes/pull/81/commits/4d9727f79b7fdc467be7bda7b7dc248a80b470e2 is the one patch which may be confusing, as I got carried away and it was getting late.

luckysori committed:
Add interval module

luckysori committed:
Handle multiple (message, nonce) pairs

luckysori committed:
Construct payouts from intervals

luckysori committed:
Introduce oracle module

luckysori committed:
Remove rand_chacha

We don't need to use seedable RNG in our tests.

luckysori committed:
Simplify test oracle attest API

thomaseizinger commented on cfd_protocol/src/interval.rs:
```diff
@@ -0,0 +1,124 @@
+/// Maximum number of binary digits for BTC price in whole USD.
+const MAX_DIGITS: usize = 20;
+
+const MAX_PRICE_DEC: u64 = (BASE as u64).pow(MAX_DIGITS as u32);
```
`const` :heart_eyes:

thomaseizinger commented on cfd_protocol/src/interval.rs:
```diff
@@ -0,0 +1,124 @@
+    use super::*;
+    use proptest::prelude::*;
+
+    impl PartialEq<Vec<Digits>> for Interval {
+        fn eq(&self, other: &Vec<Digits>) -> bool {
+            let sub_intervals = other.iter().flat_map(|i| i.range());
+            sub_intervals.eq(self.0.clone())
+        }
+    }
```
Wouldn't it be clearer if that were part of the actual test?

thomaseizinger commented on cfd_protocol/src/interval.rs:
```diff
@@ -0,0 +1,124 @@
+        Ok(Self(start..=end))
+    }
+
+    pub fn as_digits(&self) -> Vec<Digits> {
```
This allocates :upside_down_face: 

```suggestion
    pub fn to_digits(&self) -> Vec<Digits> {
```

luckysori replied:
I promise I will not mess this up one day.

thomaseizinger commented on cfd_protocol/src/interval.rs:
```diff
@@ -0,0 +1,124 @@
+        })
+        .collect()
+    }
+}
```
It looks like the public API of this module is essentially: `(start, end) -> Vec<Digits>`

You could expose that as a free function and make `Interval` private (or even inline it if deemed unnecessary).

thomaseizinger commented on cfd_protocol/src/interval.rs:
```diff
@@ -0,0 +1,124 @@
+const BASE: usize = 2;
+
+/// Maximum number of binary digits for BTC price in whole USD.
+const MAX_DIGITS: usize = 20;
```
Bit more optimism Lucas!

thomaseizinger commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -833,50 +862,50 @@ where
     /// Must remain secret.
-    nonce: SecretKey,
-    nonce_pk: schnorrsig::PublicKey,
+    nonces: Vec<SecretKey>,
```
Should we make this a fixed-size array?

luckysori committed:
Prove that signing a price works

And fix bugs along the way.

luckysori committed:
Expose APIs that are needed

And move the protocol to a `protocol` sub-module.

luckysori committed:
Reorganize modules

luckysori committed:
Remove Interval from public API

luckysori committed:
Refactor cfd_protocol crate

thomaseizinger commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -877,16 +873,7 @@ impl Event {
-                (nonce, nonce_pk)
-            })
-            .unzip();
+        let (nonces, nonce_pks) = (0..20).map(|_| oracle::nonce(rng)).unzip();
```
Pure elegance.

thomaseizinger commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -265,8 +264,8 @@ fn renew_cfd() {
-fn create_cfd_txs(
-    rng: &mut ChaChaRng,
+fn create_cfd_txs<R>(
+    rng: &mut R,
```
Could use `impl RngCore + CryptoRng` to save on the `where` clause.

luckysori replied:
Did I never know that this was equivalent? Thanks!

thomaseizinger commented on cfd_protocol/tests/cfds.rs:
```diff
@@ -843,6 +920,20 @@ impl Oracle {
+        nonces: &[SecretKey; Self::MAX_DIGITS],
+    ) -> Vec<schnorrsig::Signature> {
+        let mut bits = BitVec::from_bytes(&price.to_be_bytes());
+        let bits = bits.split_off(bits.len() - 20);
```
Is this `20` related to `MAX_DIGITS`? Should use the constant for documentation purposes at least.

thomaseizinger commented on cfd_protocol/src/protocol/sighash_ext.rs:
```diff
@@ -0,0 +1,15 @@
+use bdk::bitcoin::hashes::Hash;
+use bdk::bitcoin::SigHash;
+
+pub(super) trait SigHashExt {
```
Why `pub(super)` specifically?

thomaseizinger replied:
I think `pub(crate)` would be sufficient no? (And is stable for moving things around).

luckysori replied:
Good point. I don't know what possessed me to suddenly prefer `pub(super)` over `pub(crate)`.

thomaseizinger commented on cfd_protocol/src/protocol/transactions.rs:
```diff
@@ -0,0 +1,441 @@
+use crate::protocol::{
+    commit_descriptor, compute_adaptor_point, lock_descriptor, Payout, DUMMY_2OF2_MULTISIG,
+};
+
```
Would prefer no empty lines in imports. Prevents rustfmt from re-arranging all of these.

thomaseizinger commented on cfd_protocol/src/protocol.rs:
```diff
@@ -344,12 +345,14 @@ pub struct Payout {
     pub fn new(
-        interval: Interval,
+        start: u64,
+        end: u64,
```
You could take a `Range` maybe?

thomaseizinger commented on cfd_protocol/src/protocol.rs:
```diff
@@ -0,0 +1,532 @@
+pub fn create_cfd_transactions(
+    (maker, maker_punish_params): (PartyParams, PunishParams),
+    (taker, taker_punish_params): (PartyParams, PunishParams),
+    (oracle_pk, nonce_pks): (schnorrsig::PublicKey, &[schnorrsig::PublicKey]),
```
The slice here represents the nonces per digit yes? I think we will at least need to documentation on whether that is in big or little endian order or create a data structure that encodes this somehow.

thomaseizinger reviewed (approved):
Only nits really. Thank you, was nice to review!

The implementation looks sound, I haven't checked it in super high detail but I think the APIs makes sense and I trust that the tests are good :)

luckysori committed:
Use impl Trait shorthand wherever possible

luckysori committed:
Use Oracle::MAX_DIGITS where possible in tests

luckysori committed:
Prefer pub(crate) over pub(super)

It does the same job for us and it makes it easier to move things
around later.

luckysori committed:
Remove unnecessary newlines around imports

luckysori committed:
Take RangeInclusive<u64> as argument to Payout constructor

luckysori committed:
Document create_cfd_transactions
